src/TodoApp.jsx

```diff
@@ -0,0 +1,161 @@
+import React, { useState, useEffect, useMemo, useCallback } from 'react';
+import { TodoList } from './components/TodoList';
+import { TodosFilter } from './components/TodosFilter';
+import { getfilteredTodos } from './helpers';
+
+const TodoApp = () => {
+  const [todos, setTodos] = useState([]);
+  const [newTodo, setNewTodo] = useState('');
+  const [filter, setFilter] = useState('all');
+  const completedTodos = useMemo(
+    () => todos.filter(todo => todo.completed).length,
+    [todos],
+  );
+  const activeTodos = useMemo(
+    () => todos.filter(todo => !todo.completed).length,
+    [todos],
+  );
+
+  useEffect(() => {
+    if (!localStorage.todos) {
+      localStorage.setItem('todos', JSON.stringify([]));
+    } else {
+      setTodos(JSON.parse(localStorage.getItem('todos')));
+    }
+  }, []);
+
+  useEffect(() => {
+    localStorage.setItem('todos', JSON.stringify(todos));
+  }, [todos]);
+
+  const handleSubmit = (event) => {
+    event.preventDefault();
+
+    if (!newTodo) {
+      return;
+    }
+
+    setTodos([
+      ...todos,
+      {
+        id: +new Date(),
+        title: newTodo,
+        completed: false,
+      },
+    ]);
+
+    setNewTodo('');
+  };
+
+  const updateTodoItem = useCallback(
+    (todoId, newTitle) => {
+      setTodos(todos.map((todo) => {
+        if (todo.id !== todoId) {
+          return { ...todo };
+        }
+
+        if (newTitle) {
+          return {
+            ...todo,
+            title: newTitle,
+          };
+        }
+
+        return {
+          ...todo,
+          completed: !todo.completed,
+        };
+      }));
+    },
+    [todos],
+  );
+
+  const toogleAll = (event) => {
+    setTodos(todos.map(todo => (
+      {
+        ...todo,
+        completed: event.target.checked,
+      }
+    )));
+  };
+
+  const filteredTodos = useMemo(
+    () => getfilteredTodos(todos, filter),
+    [todos, filter],
+  );
+
+  const deleteTodo = useCallback(
+    (todoId) => {
+      setTodos(todos.filter(todo => todo.id !== todoId));
+    },
+    [todos],
+  );
+
+  const clearCompleted = () => {
+    setTodos(todos.filter(todo => !todo.completed));
+  };
```

Comment:
you have filtering here, in case of the large todo list optimization might be useful